### PR TITLE
netty: Rely on ChannelFactory in NettyServer instead of dynamic classes

### PR DESCRIPTION
**Pull request by AnarSultanov**
#5649

### CODE REVIEW

File: netty/src/main/java/io/grpc/netty/Utils.java [hunk 1]
```diff
@@ -99,12 +102,13 @@
...
     } else {
       logger.log(Level.FINE, "Epoll is not available, using Nio.", getEpollUnavailabilityCause());
-      DEFAULT_SERVER_CHANNEL_TYPE = NioServerSocketChannel.class;
+      channelType = NioServerSocketChannel.class;
       DEFAULT_CLIENT_CHANNEL_TYPE = NioSocketChannel.class;
       DEFAULT_BOSS_EVENT_LOOP_GROUP = NIO_BOSS_EVENT_LOOP_GROUP;
       DEFAULT_WORKER_EVENT_LOOP_GROUP = NIO_WORKER_EVENT_LOOP_GROUP;
       EPOLL_EVENT_LOOP_GROUP_CONSTRUCTOR = null;
     }
+    DEFAULT_SERVER_CHANNEL_FACTORY =  new ReflectiveChannelFactory<>(channelType);
```
[the remark this answers is not in the document]
Comment (AnarSultanov) [reply netty/src/main/java/io/grpc/netty/Utils.java]: With last change we will avoid reflection for nio, but not for epoll. If we want to create custom channel factory for epoll, we need to add `io.netty.netty-transport-native-epoll` dependency, should we do this?

File: netty/src/main/java/io/grpc/netty/NettyServerBuilder.java [hunk 2]
```diff
@@ -142,16 +145,40 @@ public NettyServerBuilder addListenAddress(SocketAddress listenAddress) {
...
+  /**
+   * Specifies the {@link ChannelFactory} to create {@link ServerChannel} instances. This method is
+   * usually only used if the specific {@code ServerChannel} requires complex logic which requires
+   * additional information to create the {@code ServerChannel}. Otherwise, recommend to use {@link
+   * #channelType(Class)}.
+   *
+   * <p>It's an optional parameter. If the user has not provided an Channel type or ChannelFactory
+   * when the channel is built, the builder will use the default one which is static.
+   *
+   * <p>You must also provide corresponding {@link EventLoopGroup}. For example,
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyServerBuilder.java]: This paragraph needs to match `channelType()`'s.

File: netty/src/main/java/io/grpc/netty/Utils.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ejona86) [reply netty/src/main/java/io/grpc/netty/Utils.java]: Ah, yeah. That's fine. Epoll is easily considered a separate issue. I'd prefer it not to be changed now.

File: netty/src/main/java/io/grpc/netty/NettyServerBuilder.java (hunk 2 above)
Comment (ejona86) [reply netty/src/main/java/io/grpc/netty/NettyServerBuilder.java]: I went ahead and did the copy.

**Comment by ejona86** [comment]
@AnarSultanov, thank you!

**Pull request merged**